### ISSUE TITLE
Infinite recursion loop when using DynamicRelationField with embed=True

### ISSUE THREAD

**Issue by FalFire** [AltSchool/dynamic-rest#139]
I have two serializer classes: FolderSerializer and FileSerializer, see below. I want to include the full representation of any parent folder by default in subfolders, but in doing so I get an infinite recursion loop in filters.py going from line 339 -> line 405 -> line 495 -> line 339 etc. I do not have a loop in my object hierarchy, even when visiting the detail page of a folder without parent (=None) I get this issue. Setting a depth on the Meta class does not help, when changing to the reverse relation (including subfolders in parents) I get a stack overflow error in python resulting in 'Abort trap: 6'.

```python
class FolderSerializer(BaseSerializer):
    files = DynamicRelationField("FileSerializer", many=True, deferred=True)
    parentFolder = DynamicRelationField("FolderSerializer", embed=True)

    class Meta:
        model = Folder
        fields = ('id', 'parentFolder', 'name', 'files')
        read_only_fields = ('id', 'parentFolder', 'files')

class FileSerializer(BaseSerializer):
    class Meta:
        model = File
        fields = ('id',)     // Just for testing, only include id

class Folder(models.Model):
    id = models.IntegerField(primary_key=True)
    owner = models.ForeignKey("User", related_name="folders")
    parentFolder = models.ForeignKey('Folder', related_name='subfolders', blank=True, null=True)
    name = models.CharField(max_length=40, default="Unnamed folder")
```

**Comment by aleontiev** [comment]
Thanks for reporting this @FalFire; haven't come across this before. Have you tried removing `embed=True` to see if the issue is specific to embedding?

**Comment by FalFire** [comment]
Yes, that works fine but of course that does not accomplish my goal. I can for now do without the embedding but the bug itself stands.

**Comment by aleontiev** [comment]
Got it, just figuring out where to focus in on this bug. The fact that its only relevant for `embed=True` narrows it down significantly.

**Comment by ryochiji** [comment]
This may be a bug in the serializer cache. What happens if you disable `ENABLE_SERIALIZER_CACHE` in settings?

**Comment by aleontiev** [comment]
@FalFire I'm having trouble reproducing this on master; I tried to add a [new test case](https://github.com/AltSchool/dynamic-rest/compare/fix/139-embed?expand=1#diff-625efbbf35a2290861f596fee1b19838R1548) using our "Cat" model, which has a similar self-relationship ("parent" and "kittens"), and two new embedded API fields: "embed_parent" and "embed_kittens". 

I was not able to trigger an error and got the following responses back:

```
# kitten with parent
{"cat":{"embed_parent":{"id":3,"name":"Parent"},"id":4,,"name":"Kitten"}}
# parent with kittens
{"cat":{"name":"Parent","id":3,"embed_kittens":[{"id":4,,"name":"Kitten"}]}}
```

What versions of DRF/DREST/Django are you using?
